### PR TITLE
[expression] Enable use of square brackets to access map and array elements

### PR DESCRIPTION
**Pull request by nyalldawson**
Allows expressions like:

- array(1,2,3)[0] -> 1
- array(1,2,3)[2] -> 3
- array(1,2,3)[-1] -> 3 (Python style, negative indices count from end of array)
- array(1,2,3)[-3] -> 1
- map('a',1,'b',2)['a'] -> 1
- map('a',1,'b',2)['b'] -> 2

### CODE REVIEW

**Comment by nirvn** [comment]
This, is, great! There'll be an era known as the pre-square bracket, and a post-square bracket one, for our expression engine! ;)

Note: the parser allows for nested array/map, i.e. map('a','1','b',map('ba','2'))['b']['ba'] works, etc.

File: src/core/expression/qgsexpressionnodeimpl.cpp [hunk 1]
```diff
@@ -1548,3 +1548,97 @@ QString QgsExpressionNodeBinaryOperator::text() const
...
+    {
+      const QVariantList list = QgsExpressionUtils::getListValue( container, parent );
+      qlonglong pos = QgsExpressionUtils::getIntValue( index, parent );
+      if ( pos >= list.length() || pos < -list.length() )
+      {
+        return QVariant();
+      }
+      if ( pos < 0 )
+      {
+        // negative indices are from back of list
```
Comment (nirvn) [inline src/core/expression/qgsexpressionnodeimpl.cpp]: This is great, it removes the need to call array_length (which is difficult when arrays are created in-expression) to reach the last array element.

**Comment by nirvn** [comment]
Code looking good.

File: src/core/expression/qgsexpressionnodeimpl.cpp [hunk 2]
```diff
@@ -1548,3 +1548,97 @@ QString QgsExpressionNodeBinaryOperator::text() const
...
+        pos += list.length();
+      }
+
+      return list.at( pos );
+    }
+
+    default:
```
Comment (m-kuhn) [inline src/core/expression/qgsexpressionnodeimpl.cpp]:
```
case QVariant::QgsGeometry:
  if ( geometry.isMulti() )
    return geometry.partN( pos );
```

;P

**Review by m-kuhn — APPROVED** [review]
Most welcome

File: src/core/expression/qgsexpressionnodeimpl.cpp (hunk 2 above)
Comment (nirvn) [reply src/core/expression/qgsexpressionnodeimpl.cpp]: This would indeed be really nice.

**Pull request merged**